### PR TITLE
[NVPTX] Set v2i16 SETCC to Expand

### PR DESCRIPTION
**Pull request by frasercrmck**
Note that this refers to the return type of SETCC. This operation is not
legal in PTX but was assumed as such because v2i16 is declared a legal
type. We were already expanding v4i8 SETCC.

The DAGCombiner would in certain circumstances try to fold an extension
of an illegal v2i1 SETCC (because v2i1 is illegal) into a "legal" v2i16
SETCC, which we wouldn't have patterns for.

### CODE REVIEW

**Comment by justinfargnoli** [comment]
Notes: 

> Note that this field is a property of SelectionDAG rather than one of the ISA. The previous value wasn't incorrect, per se, but since NVPTX uses 1-bit predicate registers, either choice is possible.

This is correct (source: [Section 4.5 "Constants" of PTX ISA](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#constants)):
> For predicate-type data and instructions, integer constants are allowed and are interpreted as in C, i.e., zero values are False and non-zero values are True.

---

[Compiler Explorer reproducer](https://godbolt.org/z/cKM5csYx4). 

Incorrect transformation occurs during DAG optimization:
```
Initial selection DAG: %bb.0 'sext_setcc:entry'
SelectionDAG has 14 nodes:
  t0: ch,glue = EntryToken
              t4: v1i32,ch = load<(dereferenceable invariant load (s32) from `ptr addrspace(101) null`, addrspace 101)> t0, TargetExternalSymbol:i32'sext_setcc_param_0', undef:i32
            t5: i32 = extract_vector_elt t4, Constant:i32<0>
          t6: v2i16,ch = load<(load (s32) from %ir.p)> t0, t5, undef:i32
          t8: v2i16 = BUILD_VECTOR Constant:i16<0>, Constant:i16<0>
        t10: v2i1 = setcc t6, t8, seteq:ch
      t11: v2i16 = sign_extend t10
    t12: ch = NVPTXISD::StoreRetval<(store (s32), align 1)> t0, Constant:i32<0>, t11
  t13: ch = NVPTXISD::RET_GLUE t12



Optimized lowered selection DAG: %bb.0 'sext_setcc:entry'
SelectionDAG has 13 nodes:
  t0: ch,glue = EntryToken
            t4: v1i32,ch = load<(dereferenceable invariant load (s32) from `ptr addrspace(101) null`, addrspace 101)> t0, TargetExternalSymbol:i32'sext_setcc_param_0', undef:i32
          t5: i32 = extract_vector_elt t4, Constant:i32<0>
        t6: v2i16,ch = load<(load (s32) from %ir.p)> t0, t5, undef:i32
        t8: v2i16 = BUILD_VECTOR Constant:i16<0>, Constant:i16<0>
      t14: v2i16 = setcc t6, t8, seteq:ch
    t12: ch = NVPTXISD::StoreRetval<(store (s32), align 1)> t0, Constant:i32<0>, t14
  t13: ch = NVPTXISD::RET_GLUE t12
```

**Comment by justinfargnoli** [comment]
Where in the codebase is the `sext(setcc) -> setcc` DAG combine performed?

**Review by justinfargnoli — COMMENTED** [review]
Although I'm not opposed to this change for its own sake, I'm not confident it's the correct fix for this bug. 

Based on my understanding, the root cause of this issue is that `v2i16 setcc` isn't being recognized as an illegal operation and then legalized to `v2i16 sext(v2i1 setcc)`. 

If that's correct, then I think we'd want to ensure that `v2i16 setcc` is correctly legalized instead of trying to ensure that the DAG combiner never generates a `v2i16 setcc`.

**Comment by topperc** [comment]
I'll take a look at this. AVX512 on X86 is similar so let me see how we handled this.

**Comment by topperc** [comment]
So it looks like DAGCombiner sees that getSetCCResultType returns v2i1, but then sees that v2i1 type isn't legal.

Had it not been extended it looks like v2i1 setccs get scalarized by type legalization. Maybe you just need to mark v2i16 SETCC has Expand so LegalizeVectorOps will scalarize it?

**Comment by topperc** [comment]
> So it looks like DAGCombiner sees that getSetCCResultType returns v2i1, but then sees that v2i1 type isn't legal.
> 
> Had it not been extended it looks like v2i1 setccs get scalarized by type legalization. Maybe you just need to mark v2i16 SETCC has Expand so LegalizeVectorOps will scalarize it?

Actually just marking it Expand stops the DAGCombiner which checks

```
    if (N0.hasOneUse() && TLI.isOperationLegalOrCustom(ISD::SETCC, VT) &&        
        !TLI.isOperationLegalOrCustom(ISD::SETCC, SVT)) {
```

where VT is v2i16 and SVT is v2i1.

**Comment by frasercrmck** [comment]
Thanks for the help, @justinfargnoli and @topperc. I agree that that's a better approach. I've updated the PR.

I suspect ZeroOrOne might still make sense for PTX but I don't think we have enough testing to justify switching that up.

File: llvm/test/CodeGen/NVPTX/sext-setcc.ll [hunk 1]
```diff
@@ -0,0 +1,72 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=nvptx64 -mcpu=sm_80 | FileCheck %s
+; RUN: %if ptxas-11.0 %{ llc < %s -march=nvptx64 -mcpu=sm_80 -mattr=+ptx70 | %ptxas-verify -arch=sm_80 %}
+
+define <2 x i16> @sext_setcc_v2i1_to_v2i16(ptr %p) {
```
Comment (justinfargnoli) [inline llvm/test/CodeGen/NVPTX/sext-setcc.ll:5]: Note: `v2i1 setcc` is not expanded to two `i1 setcc`s. 

```
Optimized lowered selection DAG: %bb.0 'sext_setcc_v2i1_to_v2i16:entry'
SelectionDAG has 13 nodes:
  t0: ch,glue = EntryToken
            t3: i32,ch = load<(load (s32) from `ptr addrspace(101) null`, addrspace 101)> t0, TargetExternalSymbol:i32'sext_setcc_v2i1_to_v2i16_param_0', undef:i32
          t5: v2i16,ch = load<(load (s32) from %ir.p)> t0, t3, undef:i32
          t7: v2i16 = BUILD_VECTOR Constant:i16<0>, Constant:i16<0>
        t9: v2i1 = setcc t5, t7, seteq:ch
      t10: v2i16 = sign_extend t9
    t11: ch = NVPTXISD::StoreRetval<(store (s32), align 1)> t0, Constant:i32<0>, t10
  t12: ch = NVPTXISD::RET_GLUE t11



Type-legalized selection DAG: %bb.0 'sext_setcc_v2i1_to_v2i16:entry'
SelectionDAG has 18 nodes:
  t0: ch,glue = EntryToken
    t3: i32,ch = load<(load (s32) from `ptr addrspace(101) null`, addrspace 101)> t0, TargetExternalSymbol:i32'sext_setcc_v2i1_to_v2i16_param_0', undef:i32
  t5: v2i16,ch = load<(load (s32) from %ir.p)> t0, t3, undef:i32
            t28: i16 = extract_vector_elt t5, Constant:i32<0>
          t29: i1 = setcc t28, Constant:i16<0>, seteq:ch
        t30: i16 = sign_extend t29
            t25: i16 = extract_vector_elt t5, Constant:i32<1>
          t26: i1 = setcc t25, Constant:i16<0>, seteq:ch
        t27: i16 = sign_extend t26
      t31: v2i16 = BUILD_VECTOR t30, t27
    t11: ch = NVPTXISD::StoreRetval<(store (s32), align 1)> t0, Constant:i32<0>, t31
  t12: ch = NVPTXISD::RET_GLUE t11
```

**Review by justinfargnoli — APPROVED** [review]
LGTM

**Pull request merged**